### ISSUE TITLE
Keras model.fit takes long until training begins when sample_weight is provided

### ISSUE THREAD

**Issue by tocab** [tensorflow/tensorflow#48965]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.4.1
- Python version: 3.7.10
- CUDA/cuDNN version: 11.0
- GPU model and memory: 1080 Ti 12 GB

(+ also on colab)

**Describe the current behavior**
Fit function of keras Model class takes long to begin with the training when sample_weight is provided.
I already recognized the following:
- In the backend, the method `training_utils.handle_partial_sample_weights` is used which should not be necessary because the length of my samples corresponds to the length of my labels.
- It looks like that the statement `any(w is None for w in sample_weights)` takes very long. If I execute this with on my sample_weight numpy array, this only takes 154 ms. But could it be that sample_weight is transformed into a Tensor before that operation and it therefore takes that long?

**Describe the expected behavior**
It should not have a huge impact on the runtime if sample_weight are provided.

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1Q5pTFK56AlhYAXNzlRLX--oOceQJc58_?usp=sharing

**Comment by UsharaniPagadala** [comment]
@rmothukuru 

I was able to reproduce the issue error.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/0bd0c009d5cc5ad8f1f3d945dc6817ac/-48965.ipynb) here .Thanks

**Comment by yrogory** [comment]
I investigated this a bit because it has been annoying me for some time. In my understanding, sample_weights in training_utils.handle_partial_sample_weights is either 'None' or a tensor object. If it is 'None', the check is fast because it checks if it is None before doing the partial check. The second check using list comprehension is indeed very slow and can be optimized by using array operations using numpy array. See my pull request (and change the training_utils.py in your installation manually if you like).

**Comment by tilakrayal** [comment]
@tocab,
There was the PR raised for the similar issue which was approved and merged. Could you please take a look at the issue and also kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/59a04dc8ae31a935cdb2231229727c52/_48965.ipynb).
https://github.com/keras-team/keras/pull/17357
 Thank you!

**Comment by tocab** [comment]
I tested it and it works now, thank you for the information 😊

**Comment by tilakrayal** [comment]
@tocab,
Thank you for the confirmation and Glad the issue was resolved. Could you please move this issue to closed status. Thank you!
